[2.9 ACRA] crash in DeckTask.doInBackgroundRenderBrowserQA

This  might be data related?

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/79e230f5-d99c-4b9c-8518-9db2caaf9920

```
java.lang.RuntimeException: An error occurred while executing doInBackground()
at android.os.AsyncTask$3.done(AsyncTask.java:354)
at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
at java.util.concurrent.FutureTask.run(FutureTask.java:271)
at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
at java.lang.Thread.run(Thread.java:764)
Caused by: java.lang.IndexOutOfBoundsException: Index: 446, Size: 0
at java.util.ArrayList.get(ArrayList.java:437)
at com.ichi2.async.DeckTask.doInBackgroundRenderBrowserQA(DeckTask.java:9)
at com.ichi2.async.DeckTask.doInBackground(DeckTask.java:29)
at com.ichi2.async.DeckTask.doInBackground(DeckTask.java:1)
at android.os.AsyncTask$2.call(AsyncTask.java:333)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
... 4 more
``````

--- a/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java
@@ -1161,7 +1161,7 @@ public class CardBrowser extends NavigationDrawerActivity implements
         }
         if (colIsOpen() && mCardsAdapter!= null) {
             // clear the existing card list
-            getCards().clear();
+            mCards = new ArrayList<Map<String, String>>();
             mCardsAdapter.notifyDataSetChanged();
             //  estimate maximum number of cards that could be visible (assuming worst-case minimum row height of 20dp)
             int numCardsToRender = (int) Math.ceil(mCardsListView.getHeight()/

--- a/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java
@@ -81,6 +81,7 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.Iterator;
 import java.util.LinkedHashSet;
 import java.util.List;
@@ -1393,26 +1394,26 @@ public class CardBrowser extends NavigationDrawerActivity implements
      * Removes cards from view. Doesn't delete them in model (database).
      */
     private void removeNotesView(Card[] cards) {
-        List<Integer> posList = new ArrayList<>();
         long reviewerCardId = getReviewerCardId();
-        Map<Long, Integer> idToPos = getPositionMap(getCards());
+        List<Map<String, String>> oldMCards = getCards();
+        Map<Long, Integer> idToPos = getPositionMap(oldMCards);
+        Set<Long> idToRemove = new HashSet<Long>();
         for (Card card : cards) {
-            int pos = idToPos.containsKey(card.getId()) ? idToPos.get(card.getId()) : -1;
             if (card.getId() == reviewerCardId) {
                 mReloadRequired = true;
             }
-            if (pos >= 0 && pos < getCards().size()) {
-                posList.add(pos);
+            if (idToPos.containsKey(card.getId())) {
+                idToRemove.add(card.getId());
             }
         }
 
-        // sort in descending order so we can delete all
-        Collections.sort(posList, Collections.reverseOrder());
-
-        for (int delPos : posList) {
-            getCards().remove(delPos);
+        List<Map<String, String>> newMCards = new ArrayList<Map<String, String>>();
+        for (Map<String, String> card: oldMCards) {
+            if (! idToRemove.contains(Long.parseLong(card.get("id")))) {
+                newMCards.add(card);
+            }
         }
-
+        mCards = newMCards;
         updateList();
     }
 

--- a/AnkiDroid/src/main/java/com/ichi2/async/DeckTask.java
+++ b/AnkiDroid/src/main/java/com/ichi2/async/DeckTask.java
@@ -892,13 +892,16 @@ public class DeckTask extends BaseAsyncTask<DeckTask.TaskData, DeckTask.TaskData
                 Timber.d("doInBackgroundRenderBrowserQA was aborted");
                 return null;
             }
-            if (i >= 0 && i < items.size() && items.get(i).get("answer") == null) {
-                // Extract card item
-                Card c = col.getCard(Long.parseLong(items.get(i).get("id"), 10));
-                // Update item
-                CardBrowser.updateSearchItemQA(mContext, items.get(i), c);
-                float progress = (float) i / n * 100;
-                publishProgress(new TaskData((int) progress));
+            if (i >= 0 && i < items.size()) {
+                Map<String, String> card = items.get(i);
+                if (card.get("answer") == null) {
+                    // Extract card item
+                    Card c = col.getCard(Long.parseLong(card.get("id"), 10));
+                    // Update item
+                    CardBrowser.updateSearchItemQA(mContext, card, c);
+                    float progress = (float) i / n * 100;
+                    publishProgress(new TaskData((int) progress));
+                }
             }
         }
         return new TaskData(items);

--- a/AnkiDroid/src/main/java/com/ichi2/async/DeckTask.java
+++ b/AnkiDroid/src/main/java/com/ichi2/async/DeckTask.java
@@ -881,7 +881,7 @@ public class DeckTask extends BaseAsyncTask<DeckTask.TaskData, DeckTask.TaskData
     private TaskData doInBackgroundRenderBrowserQA(TaskData... params) {
         Timber.d("doInBackgroundRenderBrowserQA");
         Collection col = CollectionHelper.getInstance().getCol(mContext);
-        List<Map<String, String>> items = (List<Map<String, String>>) params[0].getObjArray()[0];
+        List<Map<String, String>> cards = (List<Map<String, String>>) params[0].getObjArray()[0];
         Integer startPos = (Integer) params[0].getObjArray()[1];
         Integer n = (Integer) params[0].getObjArray()[2];
 
@@ -892,8 +892,8 @@ public class DeckTask extends BaseAsyncTask<DeckTask.TaskData, DeckTask.TaskData
                 Timber.d("doInBackgroundRenderBrowserQA was aborted");
                 return null;
             }
-            if (i >= 0 && i < items.size()) {
-                Map<String, String> card = items.get(i);
+            if (i >= 0 && i < cards.size()) {
+                Map<String, String> card = cards.get(i);
                 if (card.get("answer") == null) {
                     // Extract card item
                     Card c = col.getCard(Long.parseLong(card.get("id"), 10));
@@ -904,7 +904,7 @@ public class DeckTask extends BaseAsyncTask<DeckTask.TaskData, DeckTask.TaskData
                 }
             }
         }
-        return new TaskData(items);
+        return new TaskData(cards);
     }
 
 

